src/lib/ares_process.c

```diff
@@ -804,116 +843,156 @@ void ares__send_query(ares_channel channel, struct query *query,
-              next_server(channel, query, now);
-              return;
-            }
-        }
-      sendreq = ares_malloc(sizeof(struct send_request));
-      if (!sendreq)
-        {
-        end_query(channel, query, ARES_ENOMEM, NULL, 0);
-          return;
-        }
-      memset(sendreq, 0, sizeof(struct send_request));
-      /* To make the common case fast, we avoid copies by using the query's
-       * tcpbuf for as long as the query is alive. In the rare case where the
-       * query ends while it's queued for transmission, then we give the
-       * sendreq its own copy of the request packet and put it in
-       * sendreq->data_storage.
-       */
-      sendreq->data_storage = NULL;
-      sendreq->data = query->tcpbuf;
-      sendreq->len = query->tcplen;
-      sendreq->owner_query = query;
-      sendreq->next = NULL;
-      if (server->qtail)
-        server->qtail->next = sendreq;
-      else
-        {
-          SOCK_STATE_CALLBACK(channel, server->tcp_socket, 1, 1);
-          server->qhead = sendreq;
-        }
-      server->qtail = sendreq;
-      query->server_info[query->server].tcp_connection_generation =
-        server->tcp_connection_generation;
-    }
-  else
-    {
-      if (server->udp_socket == ARES_SOCKET_BAD)
-        {
-          if (open_udp_socket(channel, server) == -1)
-            {
-              skip_server(channel, query, query->server);
-              next_server(channel, query, now);
-              return;
-            }
-        }
-      if (socket_write(channel, server->udp_socket, query->qbuf, query->qlen) == -1)
-        {
-          /* FIXME: Handle EAGAIN here since it likely can happen. */
-          skip_server(channel, query, query->server);
+  if (query->using_tcp) {
+    /* Make sure the TCP socket for this server is set up and queue
+     * a send request.
+     */
+    if (server->tcp_conn == NULL) {
+      int err = open_socket(channel, server, 1);
+      switch (err) {
+        /* Good result, continue on */
+        case ARES_SUCCESS:
+          break;
+
+        /* These conditions are retryable as they are server-specific
+         * error codes */
+        case ARES_ECONNREFUSED:
+        case ARES_EBADFAMILY:
+          skip_server(channel, query, server);
           next_server(channel, query, now);
           return;
-        }
+
+        /* Anything else is not retryable, likely ENOMEM */
+        default:
+          end_query(channel, query, err, NULL, 0);
+          return;
+      }
     }
 
-    /* For each trip through the entire server list, double the channel's
-     * assigned timeout, avoiding overflow.  If channel->timeout is negative,
-     * leave it as-is, even though that should be impossible here.
+    conn = server->tcp_conn;
+
+    sendreq = ares_malloc(sizeof(struct send_request));
+    if (!sendreq) {
+      end_query(channel, query, ARES_ENOMEM, NULL, 0);
+      return;
+    }
+    memset(sendreq, 0, sizeof(struct send_request));
```

Comment:
I'd agree, however I don't care for `calloc`'s classic argument pattern in general.  Should probably also support a similar form of `realloc` to zeroize the extended chunk (though it would require knowing the original size).  We could always prefix a `size_t` sized length on all buffers returned for that purpose and return the pointer to the start of the user-data (could also be used to do things like auto-zeroize on `free`), but not sure if that's worthwhile.

c-ares is really missing a lot of nice-to-haves, like a buffer builder and generic "safe" parser.  I cringe every time I look at all the hand-written parsers.